**Цель работы:**

Осуществить предварительную обработку данных csv-файла, выявить и устранить проблемы в этих данных.

# Загрузка набора данных

### Описание предметной области

Вариант № 7

Набор данных: credit_risk.csv

Атрибуты:
- id - id записи
- age - возраст
- income - доход
- home - статус домовладения
- emp_length - стаж
- intent - намерение
- amount - сумма кредита
- rate - ставка кредита
- status - статус кредита
- percent_income - соотношение суммы кредита к доходу
- default - был ли дефолт/просрочка у клиента
- cred_length - длина кредита по времени

### 1.Чтение файла (набора данных)

In [2342]:
# импорт библиотек, чтение файла с помощью pandas
import pandas as pd
import numpy as np

# считываем файл в датафрейм
df = pd.read_csv('credit_risk.csv')

### 2. Обзор данных

2.1 Вывод первых 20 строк с помощью метода head.

In [2343]:
# по дефолту выводится лишь 5, меняем значение на 20
df.head(20)

,Id,Age,Income,Home,Emp_length,Intent,Amount,Rate,Status,Percent_income,Default,Cred_length
0,0,22.0,59000,RENT,123.0,PERSONAL,35000,16.02,1,0.59,Y,3
1,1,21.0,9600,OWN,5.0,EDUCATION,1000,11.14,0,0.10,N,2
2,2,25.0,9600,MORTGAGE,1.0,MEDICAL,5500,12.87,1,0.57,N,3
3,3,23.0,65500,RENT,4.0,MEDICAL,35000,15.23,1,0.53,N,2
4,4,24.0,54400,RENT,8.0,MEDICAL,35000,14.27,1,0.55,Y,4
5,5,21.0,9900,OWN,2.0,VENTURE,2500,7.14,1,0.25,N,2
6,6,26.0,77100,RENET,8.0,EDUCATION,35000,12.42,1,0.45,N,3
7,7,24.0,78956,RENT,5.0,MEDICAL,35000,11.11,1,0.44,N,4
8,8,24.0,83000,RENT,8.0,PERSONAL,35000,8.90,1,0.42,N,2
9,9,21.0,10000,OWN,6.0,VENTURE,1600,14.74,1,0.16,N,3


2.2 Оценка данных с помощью метода info.

In [2344]:
# выполнить метод info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 652 entries, 0 to 651
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Id              652 non-null    int64  
 1   Age             652 non-null    float64
 2   Income          652 non-null    int64  
 3   Home            652 non-null    object 
 4   Emp_length      643 non-null    float64
 5   Intent          652 non-null    object 
 6   Amount          652 non-null    int64  
 7   Rate            586 non-null    float64
 8   Status          652 non-null    int64  
 9   Percent_income  652 non-null    float64
 10  Default         652 non-null    object 
 11  Cred_length     652 non-null    int64  
dtypes: float64(4), int64(5), object(3)
memory usage: 61.3+ KB


2.3 Оценка данных с помощью метода describe.

In [2345]:
# оцените числовые столбцы с помощью describe
df.describe()

,Id,Age,Income,Emp_length,Amount,Rate,Status,Percent_income,Cred_length
count,652.000000,652.000000,652.000000,643.000000,652.000000,586.000000,652.000000,652.000000,652.000000
mean,325.495399,24.285276,90008.855828,4.601866,18800.996933,12.289471,0.601227,0.279831,3.012270
std,188.352535,7.841061,69530.312525,7.281616,9152.896250,3.272772,0.490022,0.150231,0.808843
min,0.000000,21.000000,9600.000000,0.000000,1000.000000,5.420000,0.000000,0.010000,2.000000
25%,162.750000,23.000000,44000.000000,2.000000,10000.000000,10.250000,0.000000,0.160000,2.000000
50%,325.500000,24.000000,69998.000000,4.000000,21850.000000,12.180000,1.000000,0.280000,3.000000
75%,488.250000,25.000000,128499.000000,7.000000,25000.000000,14.720000,1.000000,0.380000,4.000000
max,649.000000,144.000000,500000.000000,123.000000,35000.000000,21.210000,1.000000,0.830000,4.000000



---

Выводы по head(20):
- Видно что id дублирует стандартные индексы строк, но возможно id в датасете отвечает за id конкретной записи или конкретного человека, поэтому позже нужно нужно будет обратить на это внимание.
- Видно возможную опечатку в колонке home - RENET
- Видно невалидное значение в колонке emp_length - 123.0, вряд ли человек мог работать 123 года.

Выводы по info():
- Столбец age имеет вещественный тип, хотя в head(20) все возраста имеют целочисленный тип, да и хранить возраст как вещественное число в даном контексте не имеет смысла.
- Аналогичная ситуция с вещественным типом у emp_length - тоже не имеет смысла.
- Столбец status имеет целочисленный тип, хотя в head(20) и по логике данного атрибута должен быть булевым (да/нет).
- Столбец default имеет тип object, хотя как и status по смыслу должен быть булевым.

Выводы по describe():
- Столбец id имеет максимальное значение 649, хотя записей всего 652, вероятно id дублируется или пропущено.
- Столбец age имеет масимальное значение 144 - это не логично, человек столько жить не может.
- Столбец emp_length имеет количество записей 643, значит некоторые записи пропущены.
- Столбец rate также имеет меньшее количество записей - 586.
---


 2.4 Оценка названий столбцов

In [2346]:
# Вывести на экран названия столбцов с помощью df.columns. 
# Выявить проблемы с названиями, если они есть. 
# При необходимости переименовать столбцы.
# Если проблемы не обнаружены также дать пояснения.

# Вывод названия столбцов
print('Исходные названия столбцов')
print(df.columns)

# Для удобства работы, переведу названия столбцов в snake_case и уберу сокращения.
# Дополнительно расширю некоторые термины для полного контекста названия.
df = df.rename(columns={
        'Id': 'id',
        'Age': 'age',
        'Income': 'income',
        'Home': 'home_ownership',
        'Emp_length': 'employee_length',
        'Intent': 'intent',
        'Amount': 'amount',
        'Rate': 'rate',
        'Status': 'status',
        'Percent_income': 'percent_income',
        'Default': 'default',
        'Cred_length': 'credit_length',
})

# Вывод новых названий столбцов
print()
print('Новые названия столбцов')
print(df.columns)

Исходные названия столбцов
Index(['Id', 'Age', 'Income', 'Home', 'Emp_length', 'Intent', 'Amount', 'Rate',
       'Status', 'Percent_income', 'Default', 'Cred_length'],
      dtype='object')

Новые названия столбцов
Index(['id', 'age', 'income', 'home_ownership', 'employee_length', 'intent',
       'amount', 'rate', 'status', 'percent_income', 'default',
       'credit_length'],
      dtype='object')


### 3. Проверка пропусков

In [2347]:
# Проверить данные на наличие пропусков и устранить их, если они есть 
# (пропуски необходимо либо удалить, либо заменить каким-то значением).

# Из анализа df.describe пропуски были у колонок:
# - employee_length
# - rate

na_df = df.isna().sum()
na_employee_length = df['employee_length'].isna().sum()
na_rate = df['rate'].isna().sum()

df_count = df.shape[0]

print('Общее количество пропусков по каждой колонке')
print(na_df)

print()
print('Суммарное количество пропусков в датасете')
print(na_df.sum())

print()
print('Количество пропусков в колонках employee_length и rate суммарно')
print(na_employee_length + na_rate)

print()
print('Процент пропусков в колонке employee_length')
print(na_employee_length / df_count * 100)

print()
print('Процент пропусков в колонке rate')
print(na_rate / df_count * 100)

Общее количество пропусков по каждой колонке
id                  0
age                 0
income              0
home_ownership      0
employee_length     9
intent              0
amount              0
rate               66
status              0
percent_income      0
default             0
credit_length       0
dtype: int64

Суммарное количество пропусков в датасете
75

Количество пропусков в колонках employee_length и rate суммарно
75

Процент пропусков в колонке employee_length
1.3803680981595092

Процент пропусков в колонке rate
10.122699386503067


In [2348]:
# Количество пропуском во всем датасете сошлось с количеством пропуском в колонках employee_length и rate,
# что и следовало ожидать.

# Так как процент пропусков по полю employee_length составляет всего ~1%,
# а разрброс значений не является большим (25 перцентиль - 2, а 75 перцентиль - 7)
# и дополнительно данное поле не является ключевым при анализе риска по кредитам,
# то лучшем решением будет заменить пропуски медианным значением.

df['employee_length'] = df['employee_length'].fillna(df['employee_length'].median())
print('Количество пропусков в employee_length после замены')
print(df['employee_length'].isna().sum())

# Так как количество пропусков по полю rate составляет около 11%, что довольно много, 
# а учитывая тот факт, что данное поле является важным при анализе риска по кредитам,
# то лучшем решением будет удалить строки, где rate пропущен.
df = df.dropna(subset=['rate'])
# Восстанавливаем индексацию при помощи reset_index.
# Не создаем столбец index, при помощи аргумента drop=True.
df = df.reset_index(drop=True)
print()
print('Количество пропусков в rate после удаления строк')
print(df['rate'].isna().sum())

print()
print('Количество пропусков в датасете после замены пропусков')
print(df.isna().sum().sum())

print()
print('Колечество строк в датафрейме')
print(df.shape[0])

Количество пропусков в employee_length после замены
0

Количество пропусков в rate после удаления строк
0

Количество пропусков в датасете после замены пропусков
0

Колечество строк в датафрейме
586



---

Выводы по проверке и исправлению пропусков:
- Пропуски в данных были выявлены еще при анализе вывода describe(), дополинтельно подтвердили наличие пропусков с помощью isna().sum().
- Всего две колонки содержали пропуски, но обработаны были по разному.
- Пропуски в колонке employee_length были заменены медианным значением, а строки с пропусками в rate были удалены.
- Еще одним возможным решением для исправления пропусков является заполнение данных нулевым состоянием (0, '', None) или стандартным значением, если это применимо в конкретной ситуации - в данной ситауции это неприменимо. 


---

### 4. Проверка дубликатов и аномалий в данных

#### Проверка явных дубликатов

In [2349]:
# Проверяем количество явных дубликатов при помощи duplicated
print('Количество явных дубликатов')
print(df.duplicated().sum())

Количество явных дубликатов
2


In [2350]:
# Удаление явных дубликатов при помощи drop_duplicates
df = df.drop_duplicates()

# Восстанавливаем индексацию при помощи reset_index.
# Не создаем столбец index, при помощи аргумента drop=True.
df = df.reset_index(drop=True)

print('Колечество строк в датафрейме после удаления явных дубликатов')
print(df.shape[0])

Колечество строк в датафрейме после удаления явных дубликатов
584


#### Проверка неявных дубликатов

In [2351]:
# Неявные дубликаты могут содержаться в колонках с типом object, так как могут быть написаны неверно.
# В текущем датасете это:
# - home_ownership
# - intent
# - default

# Для поиска неявных дубликатов воспользуемся методом unique для нахождения всех уникальных значений в столбце.
print('Уникальные значения в столбце home_ownership')
print(df['home_ownership'].unique())

print()
print('Уникальные значения в столбце intent')
print(df['intent'].unique())

print()
print('Уникальные значения в столбце default')
print(df['default'].unique())

# Дополнительно выведем количество каждого уникального значения в столбце.
print()
print('Количество каждого уникального значения в столбце home_ownership')
print(df['home_ownership'].value_counts())

print()
print('Количество каждого уникального значения в столбце intent')
print(df['intent'].value_counts())

print()
print('Количество каждого уникального значения в столбце default')
print(df['default'].value_counts())

Уникальные значения в столбце home_ownership
['RENT' 'OWN' 'MORTGAGE' 'RENET' 'OTHER']

Уникальные значения в столбце intent
['PERSONAL' 'EDUCATION' 'MEDICAL' 'VENTURE' 'HOMEIMPROVEMENT'
 'DEBTCONSOLIDATION']

Уникальные значения в столбце default
['Y' 'N' 'No']

Количество каждого уникального значения в столбце home_ownership
home_ownership
RENT        377
MORTGAGE    119
OWN          85
OTHER         2
RENET         1
Name: count, dtype: int64

Количество каждого уникального значения в столбце intent
intent
EDUCATION            131
VENTURE              109
PERSONAL             104
MEDICAL               92
DEBTCONSOLIDATION     88
HOMEIMPROVEMENT       60
Name: count, dtype: int64

Количество каждого уникального значения в столбце default
default
N     445
Y     136
No      3
Name: count, dtype: int64


In [2352]:
# Удаление дубликатов

# В столбце intent отсутствуют дубликаты, а количество уникальных значений intent распределено нормально.
# Никаких действий не требуется.

# В столбце home_ownership, видно явную опечатку в виде RENET.
# Также в столбце присуствет уникальное значение OTHER, но всего в 2 строках,
# Так как 2 строки не играют роли в изучении рисков по кредитам, они будут удалены.

# Переименовываем RENET в RENT
df['home_ownership'] = df['home_ownership'].replace('RENET', 'RENT')

# Удаляем строки с OTHER и восстанавливаем индексацию
df = df[df['home_ownership'] != 'OTHER']
df = df.reset_index(drop=True)

# В столбце default наблюдается уникальный вариант No, что является аналогией к N.
# Переименовываем No в N
df['default'] = df['default'].replace('No', 'N')

print('Уникальные значения в столбце home_ownership')
print(df['home_ownership'].unique())

print()
print('Уникальные значения в столбце default')
print(df['default'].unique())



Уникальные значения в столбце home_ownership
['RENT' 'OWN' 'MORTGAGE']

Уникальные значения в столбце default
['Y' 'N']


#### Проверка аномалий

In [2353]:
# Ранее в анализе были обнаружены следующие аномалии (не включая уже разрешенные)
# - слишком большое максимальное значение у employee_length
# - слишком большое максимальное значение у age
# - id дублирует стандартные индексы
# - status хоть целочисленный и по логике должен содержать 0 или 1, но возможно там есть и другие значения.

# Подготовим данные для изменений
# Найдем 5 максимальных значений employee_length,
# И если они будут нормальными, заменим аномалный employee_length на медиану.
# А если нет, то выставим барьер самостоятельно в некоторую константу.
print("5 максимальных значений employee_length")
employee_length_largest = df['employee_length'].nlargest(5)
print(employee_length_largest)

# Найдем 5 максимальных значений age по такой же логике.
print()
print("5 максимальных значений age")
age_largest = df['age'].nlargest(5)
print(age_largest)

# Посмотрим, есть ли дубликаты в id и если есть, выведем.
print()
print("Дубликаты в id")
print(df[df.duplicated('id')])

# Посмотрим на уникальные значения status
print()
print("Уникальные значения status")
print(df['status'].unique())

5 максимальных значений employee_length
0      123.0
186    123.0
125     11.0
494     11.0
33      10.0
Name: employee_length, dtype: float64

5 максимальных значений age
73     144.0
163    144.0
514    123.0
6       26.0
13      26.0
Name: age, dtype: float64

Дубликаты в id
Empty DataFrame
Columns: [id, age, income, home_ownership, employee_length, intent, amount, rate, status, percent_income, default, credit_length]
Index: []

Уникальные значения status
[1 0]


In [2354]:
# Аномалий в столбце id не обнаружено.
# Видимо в ранних обработках датасета, аномалия уже была убрана.
# Произойти это могло при удалении или фильтрации.

# Аномалий в status также не обнаружено.
# Все строки содержат сторого 1 или 0.

# Из вывода максимальных значений employee_length видно 2 выброса по значению 123.
# Заменим эти выбросы медианой.
# Замена в данном случае хороший вариант - так как мы сохраним данные, пусть и не много,
# а медиана не разрушит существующие тендендции.
# Дополнительно - стаж не такой важный параметр при расчете рисков по кредитам.
df['employee_length'] = df['employee_length'].replace(123, df['employee_length'].median())

# Из вывода по возрасту также видно 3 выброса
# - 144
# - 123
# Так как возраст также не важный параметр, заменим все значение >100 медианой.
# Чтобы не делать replace два раза, заменим значения >100 на NaN, а потом медианой.
# Замена будет производится при помощи loc.
df.loc[df['age'] > 100, 'age'] = np.nan
df['age'] = df['age'].fillna(df['age'].median())

print('После обработки')
print(df[['age', 'employee_length']].describe())

После обработки
              age  employee_length
count  582.000000       582.000000
mean    23.771478         4.257732
std      1.482315         3.013264
min     21.000000         0.000000
25%     23.000000         2.000000
50%     24.000000         4.000000
75%     25.000000         7.000000
max     26.000000        11.000000


---

Выводы по дубликатам:
- Явные дубликаты были легко удалены при помощи drop_duplicates и drop_duplicates.
- Неявные дубликаты были найдены при помощи анализа unique, а затем удалены или изменены другими значениями.
- Дополнительно была удалена немногочисленная группа OTHER, которая ничего не дала бы для анализа.

Выводы по аномалиям:
- Все необработанные аномалии заключались в больших максимальных значениях и были изменены на медиану.

---

### 5. Провека типов данных

In [2355]:
# Проверьте типы данных, при необходимости измените типы данных, чтобы они соответствовали действительности.

# Из раннего анализа были выявлены следующие проблемы с типами
# - age имеет вещественный тип данных
# - employee_length тоже имеет вещественный тип данных
# - status имеет целочисленный тип данных
# - default имеет тип object

# Исправление типов данных

# Исправление для age
df['age'] = df['age'].astype('int64')

# Исправление для employee_length
df['employee_length'] = df['employee_length'].astype('int64')

# Исправление для status
df['status'] = df['status'].astype('bool')

# Исправление для default
df['default'] = (df['default'] == 'Y')

print('Типы данных после исправления:')
changed_types = df[['age', 'employee_length', 'status', 'default']]
print(changed_types.info())
print()
print(changed_types.head())

Типы данных после исправления:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 582 entries, 0 to 581
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype
---  ------           --------------  -----
 0   age              582 non-null    int64
 1   employee_length  582 non-null    int64
 2   status           582 non-null    bool 
 3   default          582 non-null    bool 
dtypes: bool(2), int64(2)
memory usage: 10.4 KB
None

   age  employee_length  status  default
0   22                4    True     True
1   21                5   False    False
2   25                1    True    False
3   23                4    True    False
4   24                8    True     True


---

Выводы по исправлению типов данных:
- Были исправлены типы данных при помощи astype где это возможно, в случае изначального float, и замены на int, число просто отбрасывало вещественную часть.
- В случае со значениями ввида [1 0] метод astype(bool) автоматически конвертировал int значения в bool
- В случае с object значениями использовалась логическая индексация, для привединия object к bool

---

### 6. Группировка данных

#### Задание 1

Группировка - цель кредита (intent) по количеству статуса домовладения (home).

In [2356]:
# Группируем данные сразу по двум столбцам.
# Применяем size() для расчета количества.
# Применяем sort_values() для сортировки результата.
# df.groupby(['intent', 'home_ownership']).size().sort_values()

# Также возможный вариант группировки - применить value_counts к столбцу home_ownership,
# Предварительно сгруппировав данные по столбцу intent.
# Получится точно такой же результат
df.groupby('intent')['home_ownership'].value_counts()

intent             home_ownership
DEBTCONSOLIDATION  RENT              64
                   MORTGAGE          17
                   OWN                7
EDUCATION          RENT              89
                   MORTGAGE          27
                   OWN               14
HOMEIMPROVEMENT    RENT              33
                   OWN               15
                   MORTGAGE          12
MEDICAL            RENT              61
                   MORTGAGE          19
                   OWN               12
PERSONAL           RENT              65
                   OWN               20
                   MORTGAGE          18
VENTURE            RENT              66
                   MORTGAGE          26
                   OWN               17
Name: count, dtype: int64

Выводы из полученной группировки:
- Статус домовлодения RENT (съем жилья) самый популярный при любой цели кредита. Значит люди снимающие жилье, чаще всех берут кредиты.
- Меньше всего берут кредиты люди, имеющие свое жилье (OWN).
- Интересный момент заключается в цели кредита "HOMEIMPROVEMENT", логически его должны предпочитать люди со статусом OWN или MORTGAGE (ипотека), но все равно больше всех берут кредит на эту цель люди с статусом RENT (съем жилья).

#### Задание 2

Группировка - статус домовладения (home) по количеству дефолтов (default). 
Создать датафрейм. 
Переименовать столбец с количеством в "count". 
Отсортировать по убыванию столбца "count".

In [2357]:
# Группируем также по нескольким столбцам.
# Аналогично прошлому заданию применяем size()
# С помощью reset_index задаем имя столбцу с количеством.
# Сортируем по убыванию столбец с количеством.
home_by_default = df.groupby(['home_ownership', 'default']).size()
home_by_default = home_by_default.reset_index(name='count')
home_by_default = home_by_default.sort_values(by='count', ascending=False)
print('Датафрейм с количеством записей по каждой комбинации home_ownership и default')
print(home_by_default)

# Но возможно в описании задания "по количеству дефолтов" имелось ввиду только default=True,
# Поэтому выполним все то же самое, только предварительно отфильтровав данные по default=True.
home_by_default_true = df[df['default'] == True]
home_by_default_true = home_by_default_true.groupby(['home_ownership']).size()
home_by_default_true = home_by_default_true.reset_index(name='count')
home_by_default_true = home_by_default_true.sort_values(by='count', ascending=False)
print()
print('Датафрейм с количеством записей по home_ownership если default=True')
print(home_by_default_true)

Датафрейм с количеством записей по каждой комбинации home_ownership и default
  home_ownership  default  count
4           RENT    False    283
5           RENT     True     95
0       MORTGAGE    False     92
2            OWN    False     73
1       MORTGAGE     True     27
3            OWN     True     12

Датафрейм с количеством записей по home_ownership если default=True
  home_ownership  count
2           RENT     95
0       MORTGAGE     27
1            OWN     12


Выводы из полученной группировки:
- Больше всего дефолтов было у людей снимающих жилье (RENT), вероятно это логично по той причине, что и большинство кредитов берут именно люди, снимающие жилье.
- Все остальные группы имеют малое количество дефолтов.

#### Задание 3

Сводная таблица (pivot_table) - средняя сумма кредита (amount) для каждой цели кредита. Отсортировать по убыванию. Округлить до трёх знаков.

In [2358]:
# Создаем сводную таблицу.
# Индексом будет цель кредита - критерий по которому мы группируем.
# Значения - сумма кредита.
# Функция агрегации - среднее значение.
# aggfunc имеет стандартное значение mean, 
# но для лучшей ясности, явно указываю этот аргумент.
pivot = df.pivot_table(
    index='intent',
    values='amount',
    aggfunc='mean'
)
# Сортируем по сумме кредита в порядке убывания.
pivot = pivot.sort_values(by='amount', ascending=False)
# Округляем до 3 знаков после запятой.
pivot = pivot.round(3)
# Переименовываем столбец с суммой кредита в mean_amount для ясности.
pivot = pivot.rename(columns={'amount': 'mean_amount'})
pivot

,mean_amount
intent,
DEBTCONSOLIDATION,19768.466
EDUCATION,19742.500
MEDICAL,19536.685
PERSONAL,19137.621
VENTURE,18023.853
HOMEIMPROVEMENT,17732.917


Выводы из полученной таблицы:
- Наибольшую средную сумму кредита имеет цель погашения кредита (DEBTCONSOLIDATION). То есть люди берут кредит, чтобы погасить кредит. Но судить вероятно стоит по медиане, а не средней сумме, так как возможны выбросы.
- Вообще цели DEBTCONSOLIDATION, EDUCATION, MEDICAL и отчасти PERSONAL имеют почти одинковую среднюю сумму кредита, что говорит о их равной "популярности", если опять же доверять средней сумме, а не медиане.
- Наименьшую среднюю сумму кредита имеет цель улучшения дома (HOMEIMPROVEMENT).

#### Задание 4

Сводная таблица (pivot_table) - средний доход (income) для каждого статуса домовладения (home) - строки и цели кредита - столбцы. Отсортировать по возрастанию home. Округлить до двух знаков.

In [2359]:
# Индекс - статус домовладения.
# Значения - доход.
# Функция агрегации - среднее значение.
# Сроки и цели кредита - столбцы.
# Непонятно, что имелось ввиду под "строки и цели кредита", вероятно "строки" = "сроки" (credit_length)
pivot = df.pivot_table(
    index='home_ownership',
    columns=['credit_length', 'intent'],
    values='income',
    aggfunc='mean'
)
# Сортируем по индексу (по возрастанию home_ownership)
pivot = pivot.sort_index()
# Округляем до 2 знаков после запятой
pivot = pivot.round(2)
pivot


credit_length                  2                                        \
intent         DEBTCONSOLIDATION  EDUCATION HOMEIMPROVEMENT    MEDICAL   
home_ownership                                                           
MORTGAGE               118040.00  129004.27       175160.00  157450.00   
OWN                     12498.00   90666.67        13133.33   17666.67   
RENT                    86530.53   70878.57       100009.33   85057.12   

credit_length                                        3             \
intent           PERSONAL    VENTURE DEBTCONSOLIDATION  EDUCATION   
home_ownership                                                      
MORTGAGE         91801.00  156540.00         165408.62  163928.00   
OWN             102500.00   57653.33          18500.00   68000.00   
RENT             85466.35   80764.60          77218.82  100513.26   

credit_length                                                  \
intent         HOMEIMPROVEMENT   MEDICAL   PERSONAL   VENTURE   
home_ownership                                                  
MORTGAGE             182320.00  76966.67  112406.22  62364.67   
OWN                   77161.75  39097.00   34130.67  15840.00   
RENT                  82491.08  76950.00   87373.93  97296.38   

credit_length                  4                                        \
intent         DEBTCONSOLIDATION  EDUCATION HOMEIMPROVEMENT    MEDICAL   
home_ownership                                                           
MORTGAGE               100240.00  126066.67        107350.0  158960.40   
OWN                     13672.00   48200.00         99825.0  180000.00   
RENT                    96001.88   89219.66         83374.0   72189.17   

credit_length                        
intent          PERSONAL    VENTURE  
home_ownership                       
MORTGAGE        74260.20  153420.54  
OWN             56080.00   68216.67  
RENT            82836.78   90093.28

Выводы из полученной таблицы:
- Всего кредиты берут на 2, 3 и 4 года.
- Наибольшая средняя сумма за все сроки кредита у людей с ипотекой (MORTGAGE).
- Наименьшая средняя сумма за все сроки кредита у людей снимающих жилье (RENT)

### Вывод

В ходе лабораторной работы ...

### Дополнительное задание

#### Дополинтельное задание №2

Создать столбец "Категория дохода (Income)" с помощью категоризации, выделив 3 категории: низкий, средний и высокий доход. 

Фильтрацию границ категорий выбрать на основе описательной статистики данных. 

Затем построить сводную таблицу, показывающую средний и медианный доход для каждой категории в разрезе статуса домовладения (home_ownership).

In [2360]:
# Описательная статистика данных для дохода
df[['income']].describe()

,income
count,582.000000
mean,91256.127148
std,69526.075552
min,9600.000000
25%,46125.000000
50%,70000.000000
75%,130000.000000
max,500000.000000


In [2361]:
# Так как в задании сказано профильтровать категории на основе описательной статистики,
# То для 3х категории (низкий, средний, высокий) дохода возьмем следующие параметры:
# Низкий доход: 25% перцентиль (0 - 25%)
# Средний доход: 75% перцентиль (25% - 75%)
# Высокий доход: max (75% - 100%)

# Таким образом средний доход будет занимать 50% сегмента, 
# когда как низкий и высокий доход будут занимать 25%.

# Если бы не нужно было использовать описательную статистику,
# то было бы логичнее разделить на 3 ровных сегмента.
# 0-33% - низкий доход, 33-66% - средний доход, 66-100% - высокий доход
# Но в задании сказано профильтровать категории на основе описательной статистики, 
# поэтому будем использовать что дает описательная статистика.

stats = df['income'].describe()

# Выделяем пределы для каждой категории
low_income = stats['25%']
mid_income = stats['75%']
high_income = stats['max']

# Создаем новый столбец с категориями.
# Используем pd.cut для распределения диапазонов.
# Диапазоны будут начинаться с нуля, а не с low_income, что логично.
df['income_category'] = pd.cut(
    df['income'],
    bins=[0, low_income, mid_income, high_income], 
    labels=['low', 'mid', 'high'],
    include_lowest=True,
)

print('Выполненное разбиение на категории')
print(df[['income', 'income_category']])
print()
print(df.groupby('income_category', observed=True).size().sort_values())

# Создаем сводную таблицу.
# Индекс - категория дохода
# Разрез/столбец - статус домовладения.
# Значение - доход.
# Функции агрегации - среднее и медиана.
# observed = False - учитывать пропуски + 
# jupiter notebook ругается на отсутсвие флага observed в целом.
pivot = df.pivot_table(
    index='income_category',
    columns='home_ownership',
    values='income',
    aggfunc=['mean', 'median'],
    observed=False
)

pivot

Выполненное разбиение на категории
     income income_category
0     59000             mid
1      9600             low
2      9600             low
3     65500             mid
4     54400             mid
..      ...             ...
577   57000             mid
578   51012             mid
579   20004             low
580   58000             mid
581   58800             mid

[582 rows x 2 columns]

income_category
high    145
low     146
mid     291
dtype: int64


mean                                  median  \
home_ownership        MORTGAGE            OWN           RENT  MORTGAGE   
income_category                                                          
low               16830.325581   15562.573529   39716.685714   18000.0   
mid              125800.000000            NaN   74698.772414  125800.0   
high             201687.693333  210440.941176  178235.962264  192000.0   

                                     
home_ownership        OWN      RENT  
income_category                      
low               15600.0   42000.0  
mid                   NaN   70000.0  
high             186000.0  168600.0

Выводы по таблице и распределению:
- Как и предполагалось, из-за деления данных по описательной статистике (перцентилям), средняя группа занимает 50% данных, а низкая и высокая 25%.
- Люди со своим жильем не попадают в категорию "средний доход", только низкий или высокий.
- Самый высокий доход по среднему значению также у собственников жилья, но по медиане, самый высокий доход у ипотечников.
- Самый низкий доход по среднему и медианному значению также у собственников жилья.

#### Дополинтельное задание №23

Отфильтровать набор данных. 

Выбрать топ 2 цели кредита по средней сумме кредита + топ 2 статуса землевладения по медианной сумме кредита. 

Создать сводные таблицы: цель кредита, статус домовладения - медианный, средний, максимальный доход.

In [2362]:
# Группируем по цели кредита,
# вычисляем среднее значение суммы кредита для каждой цели.
# Сортируем результат по возрастанию.
# Берем два первых значения.
# Переводим индексы в python список.
top2intent = df.groupby('intent')['amount'].mean().sort_values().index[:2].to_list()

print('Топ 2 целей кредита')
print(top2intent)

# Аналогично вычисляем топ 2 статуса землевладения
top2home_ownership = df.groupby('home_ownership')['amount'].median().sort_values().index[:2].to_list()

print()
print('Топ 2 статуса землевладения')
print(top2home_ownership)
print()

# Фильтруем DataFrame по топ 2 целей кредита и статусу землевладения.
# Используем isin для проверки вхождения в список.
# Используем & для объединения условий.
filtered_df = df[
    df['intent'].isin(top2intent) &
    df['home_ownership'].isin(top2home_ownership)
]

pivot = filtered_df.pivot_table(
    index=['intent', 'home_ownership'],
    values='income',
    aggfunc=['median', 'mean', 'max'],
)
pivot

Топ 2 целей кредита
['HOMEIMPROVEMENT', 'VENTURE']

Топ 2 статуса землевладения
['OWN', 'MORTGAGE']



median           mean     max
                                  income         income  income
intent          home_ownership                                 
HOMEIMPROVEMENT MORTGAGE        187500.0  154346.666667  300000
                OWN              18000.0   76443.133333  300000
VENTURE         MORTGAGE        174233.5  144113.884615  300000
                OWN              16020.0   49083.529412  275000

Выводы по таблице:
- Статус домовлодения (OWN) имеет очень низкий доход среди отфильтрованных намерений, в сравнении с HOMEIMPROVEMENT, что на целый порядок больше в сравнении медианы - или почти в два раза больше в сравнении среднего числа.
- Также видно что топ-2 намерения почти не имееют разницы в доходах в разрезе топ-2 статусов домовладения.